File: android/src/main/java/com/wortise/ads/react/RNWortiseNativeAdViewWrapper.kt
package com.wortise.ads.react

import android.content.Context
import android.widget.FrameLayout

class RNWortiseNativeAdViewWrapper(context: Context) : FrameLayout(context) {

    var factoryId: String? = null
        set(value) {
            field = value
            tryRender()
        }

    var responseId: String? = null
        set(value) {
            field = value
            tryRender()
        }


    private fun tryRender() {
        val factoryId  = factoryId  ?: return
        val responseId = responseId ?: return

        val factory = WortiseNativeAds.getFactory(factoryId) ?: return

        val ad      = RNWortiseNativeAdLoader.loadedAds[responseId] ?: return
        val loader  = RNWortiseNativeAdLoader.adLoaders[responseId] ?: return

        removeAllViews()

        val adView = factory.createNativeAdView()

        adView.layoutParams = LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)

        addView(adView)

        loader.renderAd(adView, ad)
    }
}

File: ios/RNWortiseSdk/RNWortiseNativeAdViewWrapper.swift
import React
import UIKit
import WortiseSDK

class RNWortiseNativeAdViewWrapper: UIView {

    private var nativeAdView: WANativeAdView?


    @objc
    var factoryId: NSString? {
        didSet { tryRender() }
    }

    @objc
    var responseId: NSString? {
        didSet { tryRender() }
    }


    deinit {
        nativeAdView?.removeFromSuperview()
    }


    private func tryRender() {
        guard let factoryId  = factoryId as String?,
              let responseId = responseId as String? else {
            return
        }

        guard let factory  = WortiseNativeAds.getFactory(factoryId),
              let nativeAd = RNWortiseNativeAdLoader.sharedAds[responseId],
              let loader   = RNWortiseNativeAdLoader.sharedLoaders[responseId] else {
            return
        }

        nativeAdView?.removeFromSuperview()

        let adView = factory.createNativeAdView()

        adView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        adView.frame = bounds

        insertSubview(adView, at: 0)

        nativeAdView = adView

        loader.render(ad: nativeAd, into: adView)
    }
}
